stor/tests/test_dx.py

```diff
@@ -705,52 +828,898 @@ class TestTempUrl(DXTestCase):
+        result = folder_path.listdir()
+        self.assert_dx_lists_equal(result, [dx_p])
+
+
+class TestDownloadObjects(DXTestCase):
+    def test_local_paths(self):
+        self.setup_temporary_project()
+        self.setup_file('/temp_folder/file1.txt')
+        self.setup_file('/temp_folder/file2.txt')
+        self.setup_file('/temp_folder/folder/file3.txt')
+        dx_p = DXPath('dx://' + self.project + ':/temp_folder')
+        posix_folder_p = Path('./{test_folder}'.format(
+            test_folder=self.project))
+        self.addCleanup(posix_folder_p.rmtree)
+        r = dx_p.download_objects(self.project,
+                                  ['file1.txt',
+                                   'file2.txt',
+                                   'folder/file3.txt'])
+        self.assertEquals(r, {
+            'file1.txt': self.project + '/file1.txt',
+            'file2.txt': self.project + '/file2.txt',
+            'folder/file3.txt': self.project + '/folder/file3.txt'
+        })
+
+    def test_absolute_paths(self):
+        self.setup_temporary_project()
+        self.setup_file('/temp_folder/file1.txt')
+        self.setup_file('/temp_folder/file2.txt')
+        self.setup_file('/temp_folder/folder/file3.txt')
+        dx_p = DXPath('dx://' + self.project)
+        posix_folder_p = Path('./{test_folder}'.format(
+            test_folder=self.project))
+        self.addCleanup(posix_folder_p.rmtree)
+        r = dx_p.download_objects(self.project, [
+            'dx://' + self.project + ':/temp_folder/file1.txt',
+            'dx://' + self.project + ':/temp_folder/file2.txt',
+            'dx://' + self.project + ':/temp_folder/folder/file3.txt'])
+        self.assertEquals(r, {
+            'dx://' + self.project + ':/temp_folder/file1.txt':
+                self.project + '/temp_folder/file1.txt',
+            'dx://' + self.project + ':/temp_folder/file2.txt':
+                self.project + '/temp_folder/file2.txt',
+            'dx://' + self.project + ':/temp_folder/folder/file3.txt':
+                self.project + '/temp_folder/folder/file3.txt'
+        })
+
+    def test_absolute_paths_not_child_of_download_path(self):
+        dx_p = DXPath('dx://project:/folder')
+        with pytest.raises(ValueError, match='child'):
+            dx_p.download_objects('output_dir', [
+                'dx://project:/bad/e/f.txt',
+                'dx://project:/bad/e/f/g.txt'
+            ])
+
+
+class TestCopy(DXTestCase):
+    def test_clone_move_project_fail(self):
+        self.setup_temporary_project()
+        self.setup_file('/folder_file.txt')
+        dx_p = DXPath('dx://' + self.project + ':/folder_file.txt')
+        proj_handler = dxpy.DXProject()
+        proj_handler.new('test_copy_project_fail.TempProj')
+        self.addCleanup(proj_handler.destroy)
+        proj_p = DXPath('dx://test_copy_project_fail.TempProj:/')
+        with pytest.raises(ValueError, match='Cannot clone project'):
```

Comment:
I'm a litlte confused, wouldn't 

```
stor cp -r dx://source_project:/ dx://test_project:/
```
cause all the top level files and folders of source_project to be cloned to test_project?

moving within project is a little more complicated when parent is all folders, so okay to skip that

Reply:
It would. This is a test case for copy, not copytree